src/util/LazyJsonParser.cpp

```diff
@@ -0,0 +1,258 @@
+    : arrayPath_(std::move(arrayPath)),
+      prefixInArray_(absl::StrCat(
+          absl::StrJoin(arrayPath_.begin(), arrayPath_.end(), "",
+                        [](std::string* out, const std::string& s) {
+                          absl::StrAppend(out, "{\"", s, "\": ");
+                        }),
+          "[")),
+      suffixInArray_(absl::StrCat("]", std::string(arrayPath_.size(), '}'))) {}
+
+// ____________________________________________________________________________
+std::string LazyJsonParser::parseChunk(std::string_view inStr) {
+  size_t idx = input_.size();
+  absl::StrAppend(&input_, inStr);
+
+  // End-index (exclusive) of the current `input_` to construct a result.
+  size_t materializeEnd = 0;
+
+  // If the previous chunk ended within a Literal, finish parsing it.
+  if (inLiteral_) {
+    parseLiteral(idx);
+    ++idx;
+  }
+
+  // Resume parsing the current section.
+  while (idx < input_.size()) {
+    switch (state_.index()) {
+      case 0:
+        parseBeforeArrayPath(idx);
+        break;
+      case 1:
+        parseInArrayPath(idx, materializeEnd);
+        if (arrayPath_.empty() &&
+            std::holds_alternative<AfterArrayPath>(state_)) {
+          materializeEnd = idx;
+        }
+        break;
+      case 2:
+        parseAfterArrayPath(idx, materializeEnd);
+        break;
+    }
+  }
+
+  return constructResultFromParsedChunk(materializeEnd);
+}
+
+// ____________________________________________________________________________
+void LazyJsonParser::parseLiteral(size_t& idx) {
+  AD_CORRECTNESS_CHECK(inLiteral_ || input_[idx] == '"');
+  if (input_[idx] == '"' && !inLiteral_) {
+    ++idx;
+    if (std::holds_alternative<BeforeArrayPath>(state_)) {
+      std::get<BeforeArrayPath>(state_).litStart_ = idx;
+    }
+    inLiteral_ = true;
+  }
+
+  for (; idx < input_.size(); ++idx) {
+    if (idx > 1'000'000) {
+      throw std::runtime_error(
+          "Ill-formed JSON: Element size exceeds 1 Megabyte.");
+    }
+    if (isEscaped_) {
+      isEscaped_ = false;
+      continue;
+    }
+    switch (input_[idx]) {
+      case '"':
+        // End of literal.
+        if (std::holds_alternative<BeforeArrayPath>(state_)) {
+          std::get<BeforeArrayPath>(state_).litLength_ =
+              idx - std::get<BeforeArrayPath>(state_).litStart_;
+        }
+        inLiteral_ = false;
+        return;
+      case '\\':
+        isEscaped_ = true;
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+// ____________________________________________________________________________
+void LazyJsonParser::parseBeforeArrayPath(size_t& idx) {
+  AD_CORRECTNESS_CHECK(std::holds_alternative<BeforeArrayPath>(state_));
+  auto& state = std::get<BeforeArrayPath>(state_);
+
+  for (; idx < input_.size(); ++idx) {
+    if (idx >= 1'000'000) {
+      throw std::runtime_error(
+          "Ill-formed JSON: Header size exceeds 1 Megabyte.");
+    }
```

Comment:
This mechanism only catches the case
if the single chunks are very big. 
That however is not the big problem, because that is the user's responsibility.
The problem is rather, if we parse many chunks without seeing the end of the header, or the end of an array.

The solution to this is basically also simple:
After each chunk (can be done in the very outer loop) Assert that the size of the remaining input (that you have to carry on to the next chunk as a prefix) is smaller than the given threshold if that makes sense.
So no need to add something within the single chunks, if someone wants to pass in megabytes here, this is not our problem:)